### PR TITLE
Add full repo storage

### PR DESCRIPTION
**Pull request by zangell44**
## Summary
This PR adds a generalized git storage class to Prefect.

## Changes

- Adds a new `Git` storage class which uses a pure python git implemenation
- Adds a new `TemporaryGitRepo` utility, allowing us to temporary clone and cleanup a repository

## Importance
These changes will
- Allow users to use Deploy Tokens and other ssh based git auth if they wish
- Allow us to become less reliant on third party libraries
- Allow users to pull info from non-python files at run time (python modules are another beast)

These changes address
- https://github.com/PrefectHQ/prefect/issues/4376
- https://github.com/PrefectHQ/prefect/issues/4328

These changes WILL NOT
- Allow users to import Python modules from the repo when cloning

## Example usage
```python
import prefect
from prefect import task, Flow
from prefect.storage import Git

storage = Git(flow_path='flow.py', repo='zangell44/single-prefect-flow', repo_host='github.com')

@task
def say_hello(name):
        logger = prefect.context.get("logger")
        logger.info('Hi' + name)

with Flow('my-hello-flow') as flow:
        say_hello('Zach')

flow.storage = storage
```

## Checklist

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

File: src/prefect/storage/git.py
Comment (madkinsz) [inline src/prefect/storage/git.py:59]: Can I use a commit #?

File: src/prefect/storage/git.py
Comment (madkinsz) [inline src/prefect/storage/git.py:45]:
```suggestion
        - format_access_token (bool): if True, the class will attempt to format access tokens
```

File: src/prefect/storage/git.py
Comment (madkinsz) [inline src/prefect/storage/git.py:77]: This needs a comment at least

File: src/prefect/storage/git.py
Comment (madkinsz) [inline src/prefect/storage/git.py:148]: Should probably warn if a secret is set here

File: src/prefect/storage/git.py
Comment (madkinsz) [inline src/prefect/storage/git.py:129]: Should probably support loading from an environment variable too so they don't need to use Prefect secrets to set this

**Review by madkinsz — CHANGES_REQUESTED** [review]
Good start! Will definitely need to be added to the docs

File: src/prefect/storage/git.py
Comment (zangell44) [reply src/prefect/storage/git.py:59]: not right now as implemented, I can do some more digging in `dulwich` to see what's possible here

is commit # a common enough pattern that it's must-have?

File: src/prefect/storage/git.py
Comment (zangell44) [reply src/prefect/storage/git.py:15]: oof this looked ugly in the docs

File: src/prefect/storage/git.py
Comment (zangell44) [reply src/prefect/storage/git.py:140]: Hmm are there other good examples of how we'd want this pattern to work?

My cheeky response is, "just set the environment variable `export PREFECT__CONTEXT__SECRETS__MYSECRET="MY SECRET VALUE"`:".

File: src/prefect/storage/git.py
Comment (zangell44) [reply src/prefect/storage/git.py:165]: added a warning in __init__, hopefully will allow users to catch it right away

File: src/prefect/storage/git.py
Comment (madkinsz) [reply src/prefect/storage/git.py:66]: I guess every git tool I've worked with takes branch/commit hash/tag indiscriminately and I expect the same. If it's at all a pain I don't think it should hold this up but I'd make a follow up issue.

File: src/prefect/storage/git.py
Comment (madkinsz) [reply src/prefect/storage/git.py:146]: Idk how strongly I feel but the pattern used in all the other storage's is like so
```python
        if self.access_token_secret is not None:
            # If access token secret specified, load it
            access_token = Secret(self.access_token_secret).get()
        else:
            # Otherwise, fallback to loading from local secret or environment variable
            access_token = prefect.context.get("secrets", {}).get("GITHUB_ACCESS_TOKEN")
            if access_token is None:
                access_token = os.getenv("GITHUB_ACCESS_TOKEN")
 ```

cc @cicdw @dylanbhughes on continuing to support this pattern

Comment (zangell44) [reply src/prefect/storage/git.py:146]: I would vote to stop supporting this pattern

Comment (dylanbhughes) [reply src/prefect/storage/git.py:146]: I defer to @cicdw. My feedback is that the user shouldn't have to stick with a particular name for a Prefect secret so that a single tenant can have more than one secret for GitHub storage and that's well-documented. As long as that's achieved I'm 👍

File: src/prefect/storage/git.py
Comment (zangell44) [reply src/prefect/storage/git.py:68]: turned out to be easy enough, just added

File: src/prefect/storage/git.py
Comment (zangell44) [reply src/prefect/storage/git.py:171]: Fair point, I went back and forth on building url from params vs. letting the user provide their own url.
Is there a repo configuration you've worked with where this would run into compatibility issues?

Comment (madkinsz) [reply src/prefect/storage/git.py:171]: I think having an option to provide a custom url may be reasonable but not strictly necessary in this first patch. We want this to be easy for people to use, we've got a lot of users that are data scientists.

Comment (sm-Fifteen) [reply src/prefect/storage/git.py:171]: Well, if that's any better an argument, having to specify the URL in full would also make the `use_ssh` parameter redundant, since it would now be specified in the URL's protocol rather than added by the storage module.

Comment (zangell44) [reply src/prefect/storage/git.py:171]: I agree with Michael here, I think this is reasonable to release and expand to include explicit URLs if necessary.

Thank you for taking a look at this PR in detail and detailed comments on the issue thread!

File: src/prefect/utilities/git.py
Comment (zangell44) [reply src/prefect/utilities/git.py:8]: Another fair point. Again, it's reasonable to release this as is and make updates as necessary if users run into performance issues.

I'll keep an eye out for any reports. Thanks for the detail here!

File: docs/orchestration/flow_config/storage.md
Comment (madkinsz) [inline docs/orchestration/flow_config/storage.md:253]: Add a note on how this differs from other git based storages (ie how it basically supersedes them).

File: docs/orchestration/flow_config/storage.md
Comment (madkinsz) [reply docs/orchestration/flow_config/storage.md:264]: We should have a basic example of using a deploy key. Even something like

> To use a private repository, you'll have to authenticate. We recommend using a deploy key (link to git docs?) stored as a Prefect Secret which you'll reference in the storage initialization

Also, have you tested using a secret in Cloud? I feel like sometimes you have to include `secrets=["MY_GIT_ACCESS_TOKEN"]` in the storage init call which I feel like is a bad ux and we should fix if it's required here.

Comment (zangell44) [reply docs/orchestration/flow_config/storage.md:264]: On deploy keys, we can link to the docs here https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys. 

As far as I can tell, they're just slightly enhanced ssh tokens. Maybe worth noting in the docs the two are equivalent and to set `use_ssh = True`?

Comment (zangell44) [reply docs/orchestration/flow_config/storage.md:266]: confirmed this works with Cloud Secrets
`LocalAgent` + `use_local_secrets=false` + using a secret that doesn't exist locally

Comment (zangell44) [reply docs/orchestration/flow_config/storage.md:266]: docs also updated accordingly

File: docs/orchestration/flow_config/storage.md
Comment (madkinsz) [inline docs/orchestration/flow_config/storage.md:280]: Does this mean that if I use something like a K8s agent I have to setup my deploy key in my agent container? Why can't I use a Prefect Secret for that value?

Comment (zangell44) [reply docs/orchestration/flow_config/storage.md:280]: Yes you would have to configure it in the container (or use access tokens as a Secret).

Using a Secret for that value would involve manipulation of ssh configs, which I think is outside of Prefect's scope

Comment (madkinsz) [reply docs/orchestration/flow_config/storage.md:280]: Ah so the confusing part here is gitlab has deploy _tokens_

https://docs.gitlab.com/ee/user/project/deploy_tokens/#git-clone-a-repository

Comment (zangell44) [reply docs/orchestration/flow_config/storage.md:280]: Haha yes I was confused about this as well. Deploy Tokens will work like other access tokens without issues

Comment (madkinsz) [reply docs/orchestration/flow_config/storage.md:280]: If you don't add a note about that we're going to get some questions haha, just this one last thing and this is good to merge :D

Comment (zangell44) [reply docs/orchestration/flow_config/storage.md:280]: Added a note and example here https://github.com/PrefectHQ/prefect/pull/4418/commits/9ada00c0950984503385fb211a2a297750b587f2#diff-357fea1f44b9d331a93957d9d6b6408d2ffc07866872dfff6f91f2bb7af3a83cR288

**Review by madkinsz — APPROVED** [review]

**Comment by mahiki** [comment]
Thus is a really helpful improvement

**Comment by zangell44** [comment]
Glad to hear @mahiki! Please let us know if there's anything we can do to improve it.

**Comment by jasonprado** [comment]
Hiya, I'm fairly new to Prefect. Can you explain why "Allow users to import Python modules from the repo when cloning" is not part of this–is it due to some structural way imports are handled inside the executor, or is it a non-goal in terms of design? I suspect the former as you said "python modules are another beast".

I have been surprised that getting script-based storage deployed is so easy but trying to share code between flows required deploying a custom docker image.

**Comment by madkinsz** [comment]
Hey @jasonprado -- due to the myriad of execution environments we support, it's a more complex problem to ensure the rest of your code is accessible alongside your flow. This is something we want to address eventually, but running your flow in a container is the best way to have control over its environment. I think can use `EXTRA_PIP_PACKAGES` (see https://github.com/PrefectHQ/prefect/pull/4013) to install your module in the container still e.g. 

```python
flow.run_config = DockerRun(env=dict(EXTRA_PIP_PACKAGES="git+https://github.com/USER/REPO@COMMIT#egg=MODULE"))
```

I haven't given this a test but it should do what you want. Generalizing this pattern into this storage mode is something we're interested in pursuing once we understand the best way to support it across our deployment methods.

**Comment by willianchan** [comment]
i am getting the following error using the Git Storage module:

`dulwich.client.HTTPUnauthorized: No valid credentials provided`

using the following code:

`flow.storage = Git(
    repo="GRG_GIT_CD/analytics/workflow-registration",
    flow_path="flow-test-clone-repo.py",
    repo_host="https://gitcorp.prod.cloud.ihf"
)`

i am using gitlab to store my flow, also, tried passing git_token_secret_name and git_token_username, am i doing something wrong?

**Comment by al-yap** [comment]
@willianchan wondering if you have managed to resolve your issue above? I am currently troubleshooting the same error..

**Comment by tyndyll** [comment]
> `flow.storage = Git( repo="GRG_GIT_CD/analytics/workflow-registration", flow_path="flow-test-clone-repo.py", repo_host="https://gitcorp.prod.cloud.ihf" )`

If GRG_GIT_CD is a URL it should probably be added as `repo_host`

**Pull request merged**
